### ISSUE TITLE
Zoom perfomance with layer type symbol

### ISSUE THREAD

**Issue by just-website** [maplibre/maplibre-gl-js#3084]
I'm trying to draw 2000 points on a type symbol layer and get big performance dips at every zoom step.  The"text-allow-overlap" parameter does not cause visible changes.
Now I can see that the findMatches method takes a very long time to complete at each zoom step.

Are such performance dips to be expected with the symbol type layer?
What can be done to fix it on large layers symbol type than 30k points (is it possible?)?

![image](https://github.com/maplibre/maplibre-gl-js/assets/20595147/7d2a9d98-87b5-4e54-9de2-bd0adfa35023)

https://github.com/maplibre/maplibre-gl-js/blame/18dcac916ccc0b7e7f1160533e5db5c34c28fffe/src/symbol/cross_tile_symbol_index.ts#L102

**Comment by HarelM** [comment]
Can you link to an example that demonstrates the issue? Jsbin?

**Comment by just-website** [comment]
https://codepen.io/Gleb-Shchukin/pen/QWzOLPN?editors=1010
Aactually the problem was solved recently in mapbox. But I see that it is not possible to copy their solution directly (and that's why I’m not even sure if it makes sense to add a link to the PR here).
I reproduced their PR locally and got *10 performance for the findMatch method.

**Comment by HarelM** [comment]
Yes, wet can't copy their code, you shouldn't link to their PR...

**Comment by just-website** [comment]
Do you have any ideas to fix this?

**Comment by HarelM** [comment]
There's a need to run performance analysis and better understand what's causing this.
I have no idea how to fix this.
If this is something that you need solved, I think your best approach would be to try and solve this yourself and submit a PR.
Make sure you don't look at the solution in mapbox so that you don't infringe their copyright.

**Comment by bradymadden97** [comment]
PR'ed a potential improvement here: https://github.com/maplibre/maplibre-gl-js/pull/6641
